### PR TITLE
Bh/submissions accept decline controls

### LINKED ISSUE #1683: Add Submission accept/decline controls to `SubmissionPubRow`
The `SubmissionPubRow` component from #1682 needs controls to accept/decline a submission. We've modeled these as thumbs up/down buttons. These should launch a new `Dialog` with the following controls:

- A text area to edit the email that will be sent to the submitter (using the fields #1681)
- A button to commit the acceptance/rejection and send the email
- A button that does the same thing but skips sending the email
- A cancel button

We can use the `DialogLauncher` component to launch these dialogs when the buttons are clicked.

The "thumbs up/down" buttons should not be visible on Pubs that have already been accepted or declined.

### PR DESCRIPTION
**Pull request by howebenhowe**
Resolves #1683

This PR adds approve/decline/delete controls to submission overview rows and invites editing/preventing emails to be sent to author(s) of submissions.

_Test plan_:

1. create three new submissions, for example for the collection "A Long Collection Name That Everyone Regrets Choosing!"
1. visit that collections Submissions overview
1. mark one of your new, `pending` status submissions as accepted, augment the email copy with some of your own thoughts, then submit
1. mark another one as declined (thumbs-down) and augment the copy
1. delete the third one

Currently, deleted pubs don't disappear until the page is refreshed. Other changes are stored temporarily in UI until a page refresh syncs back/front end anew.

### CODE REVIEW

File: client/containers/DashboardSubmissions/VerdictDialog.tsx [hunk 1]
```diff
@@ -0,0 +1,154 @@
...
+	initialEmailText?: DocJson;
+};
+
+const PreSubmissionBody = (props: PreSubmissionBodyProps) => {
+	const [shouldSendEmail, setShouldSendEmail] = useState(!!props.initialEmailText);
+	const [customEmailText, setCustomEmailText] = useState(props.initialEmailText || getEmptyDoc());
+	return (
+		<>
+			<div className={Classes.DIALOG_BODY}>
+				<p>Would you like to {props.actionTitle} this submission?</p>
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/VerdictDialog.tsx:38]: `props.actionTitle.toLowerCase()` so the action doesn't appear capitalized here.

File: client/containers/DashboardSubmissions/arbitrationMenu.scss [hunk 2]
```diff
@@ -0,0 +1,6 @@
+.arbitration-menu {
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/arbitrationMenu.scss:1]: This should be `arbitration-menu-component`

File: client/containers/DashboardSubmissions/verdictDialog.scss [hunk 3]
```diff
@@ -0,0 +1,24 @@
+.verdict-dialog {
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/verdictDialog.scss:1]: This should be `verdict-dialog-component`

File: client/containers/DashboardSubmissions/arbitrationMenu.scss [hunk 4]
```diff
@@ -0,0 +1,6 @@
...
+	display: grid;
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/arbitrationMenu.scss:2]: I wonder whether a:

```css
display: flex;
> *:not(:last-child) {
    margin-right: 40px;
}
```
might do the job better here. That way you don't have to specify the `gridColumn` in React, and it scales to support more (or fewer) buttons, which eliminates this kind of goofy state:

image

File: client/containers/DashboardSubmissions/VerdictDialog.tsx [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+			{submissionError ? (
+				<Callout intent="warning" title="There was an error creating this submission." />
+			) : updatedSubmission ? (
+				<PostSubmitBody
+					isHandlingSubmission={isHandlingSubmission}
+					completedName={props.completedName}
+					onClose={props.onClose}
+				/>
+			) : (
+				<PreSubmissionBody
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/VerdictDialog.tsx:142]: Nit: if this is called `PreSubmissionBody`, I think the other component should be called `PostSubmissionBody`. I also wonder if a better noun than "submission" could be used here to refer to the act of submitting the form — it's uniquely ambiguous here because we're dealing with `Submission` objects as well :)

File: client/containers/DashboardSubmissions/VerdictDialog.tsx [hunk 6]
```diff
@@ -0,0 +1,154 @@
...
+			lazy={true}
+			title={props.actionTitle}
+			className="verdict-dialog"
+			isOpen={props.isOpen}
+			onClose={props.onClose}
+		>
+			{submissionError ? (
+				<Callout intent="warning" title="There was an error creating this submission." />
+			) : updatedSubmission ? (
+				<PostSubmitBody
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/VerdictDialog.tsx:136]: I can't get this to ever show up in testing. Under what circumstances does it appear?

File: client/containers/DashboardSubmissions/VerdictDialog.tsx [hunk 7]
```diff
@@ -0,0 +1,154 @@
...
+	};
+	return (
+		<Dialog
+			lazy={true}
+			title={props.actionTitle}
+			className="verdict-dialog"
+			isOpen={props.isOpen}
+			onClose={props.onClose}
+		>
+			{submissionError ? (
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/VerdictDialog.tsx:133]: Would you consider a `renderInner()` function with a set of `if` statements? The nested ternaries are a little hard to follow here.

File: client/containers/DashboardSubmissions/VerdictDialog.tsx [hunk 8]
```diff
@@ -0,0 +1,154 @@
...
+	);
+};
+
+const VerdictDialog = (props: Props) => {
+	const [isHandlingSubmission, setIsHandlingSubmission] = useState(false);
+	const [updatedSubmission, setUpdatedSubmission] = useState(null);
+	const [submissionError, setSubmissionError] = useState(null);
+	const handleSubmission = (customEmailText: DocJson, shouldSendEmail: boolean) => {
+		setIsHandlingSubmission(true);
+		apiFetch('/api/submissions', {
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/VerdictDialog.tsx:106]: In the event you get rid of the `delete` option, you can use the `apiFetch.put` method (and then you don't have to stringify the body).

File: client/containers/DashboardSubmissions/verdictDialog.scss [hunk 9]
```diff
@@ -0,0 +1,24 @@
...
+	.email-text-header {
+		text-transform: uppercase;
+		font-size: 12px;
+		color: #888;
+		margin-bottom: 5px;
+	}
+	.minimal-editor-component {
+			height: 150px;
+			overflow: auto;
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/verdictDialog.scss:10]: Do you happen to remember what problems the `overflow: auto`, `min-height: 100%`, and `white-space: break-spaces` solve? I have been trying for years to get this component to Do The Right Thing by default, and if you're having CSS frustrations around it that may be a sign that I should tweak something internally.

File: client/containers/DashboardSubmissions/VerdictDialog.tsx [hunk 10]
```diff
@@ -0,0 +1,154 @@
...
+type Props = {
+	isOpen: boolean;
+	onClose: () => unknown;
+	actionTitle: string;
+	completedName: string;
+	apiMethod: string;
+	status?: string;
+	initialEmailText?: DocJson;
+	pub: DefinitelyHas<Pub, 'submission'>;
+	onJudgePub: (pubId: string, status?: string) => void;
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/VerdictDialog.tsx:21]: `status` can be narrowed to `SubmissionStatus`

File: client/containers/DashboardSubmissions/VerdictDialog.tsx [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+	const [customEmailText, setCustomEmailText] = useState(props.initialEmailText || getEmptyDoc());
+	return (
+		<>
+			<div className={Classes.DIALOG_BODY}>
+				<p>Would you like to {props.actionTitle} this submission?</p>
+				<p className="email-text-header">
+					<Icon icon="manually-entered-data" iconSize={12} />
+					{'  '}Email to Authors
+				</p>
+				<WorkflowTextEditor
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/VerdictDialog.tsx:43]: This doesn't make sense for the delete case (we won't ever send an email when the submission is deleted)

File: client/containers/DashboardSubmissions/SubmissionItems.tsx [hunk 12]
```diff
@@ -68,39 +85,31 @@ const SubmissionItems = (props: Props) => {
 		},
 	});
 
+	const [localStats, setLocalStats]: [{ [pubId: string]: SubmissionStatus }, any] = useState({});
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/SubmissionItems.tsx:88]: Slightly cleaner way to do this:

```tsx
const [localStats, setLocalStats] = useState<{ [pubId: string: SubmissionStatus }>({});
```
this will properly type both items in the `[value, setter]` array.

Comment (idreyn) [reply client/containers/DashboardSubmissions/SubmissionItems.tsx:88]: Could we lengthen these names to `localStatuses`? "stats" makes me think it's "statistics"

File: client/containers/DashboardSubmissions/SubmissionItems.tsx [hunk 13]
```diff
@@ -68,39 +85,31 @@ const SubmissionItems = (props: Props) => {
...
-	const renderPubs = () => {
-		return pubs.map((pub) => (
-			<PubOverviewRow
-				pub={pub}
-				key={pub.id}
-				leftIconElement="manually-entered-data"
-				hasSubmission={true}
-			/>
-		));
+	const trackLocalStatus = (pubId: string, status?: string) => {
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/SubmissionItems.tsx:96]: Perhaps `updateLocalStatus` would be a better name — I don't know whether "track" is a read-ish or write-ish operation from a glance. Also, this can be wrapped in a `useCallback()` and `status` could be narrowed to `SubmissionStatus`.

File: client/containers/DashboardSubmissions/SubmissionItems.tsx [hunk 14]
```diff
@@ -112,9 +121,30 @@ const SubmissionItems = (props: Props) => {
 				filters={overviewSearchFilters}
 			/>
 			<OverviewRows>
-				{renderPubs()}
+				{(pubs as DefinitelyHas<Pub, 'submission'>[])
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/SubmissionItems.tsx:124]: Could we make `DefinitelyHas<Pub, 'submission'>[]` the type of the `pubs` prop?

**Review by idreyn — CHANGES_REQUESTED** [review]
Looking well on its way! The `VerdictDialog` is feeling really nice to use. I'm primarily concerned with this "delete submission" button — I don't think it needs to exist, and it simplifies a lot of the code if it's removed.

**Comment by howebenhowe** [comment]
@idreyn re. deleting pubs, i was operating from a combination of the card and the [final figma view](https://www.figma.com/file/ERHvqJqCVBZ7dH2ZQBGlp8/SPUB-vision?node-id=223%3A1942), hence the delete functionality. Happy to eliminate it--we sure it's not a concern that the `declined` filter view will get clogged with accidental/spam submissions?

File: client/containers/DashboardSubmissions/SubmissionItems.tsx (hunk 14 above)
Comment (howebenhowe) [reply client/containers/DashboardSubmissions/SubmissionItems.tsx:124]: don't think so? We're pulling it out of both `useManyPubs` and `props.initialPubs`. I'm down to put it to `props.initialPubs`, but I don't understand how we could expect presence of submissions on pubs returned from `usemanyPubs`

Comment (howebenhowe) [reply client/containers/DashboardSubmissions/SubmissionItems.tsx:124]: nevermind, TS ain't complaining so neither am I.

**Comment by idreyn** [comment]
> @idreyn re. deleting pubs, i was operating from a combination of the card and the [final figma view](https://www.figma.com/file/ERHvqJqCVBZ7dH2ZQBGlp8/SPUB-vision?node-id=223%3A1942), hence the delete functionality. Happy to eliminate it--we sure it's not a concern that the `declined` filter view will get clogged with accidental/spam submissions?

Yes, I think this is just a miscommunication in design/dev handoff. I don't know what @djagdish's intention for the delete button was, but my guess is that it should delete the Pub itself and not just the Submission object. It would be annoying for the person managing the Submissions queue to find a Pub that was so odious that they press the "delete" button rather than just declining it, and then later find out that that the Pub is still bumping around their Community. At any rate, I apologize for not accounting for it in #1683.

File: client/containers/DashboardSubmissions/VerdictDialog.tsx (hunk 6 above)
Comment (howebenhowe) [reply client/containers/DashboardSubmissions/VerdictDialog.tsx:137]: Good catch. Got some logic conflict here. You can get it to appear by starting in the 'all section' and then either approving or declining a submission. The conflict at play is that after changing a status, rather than letting the row languish in a filter view that ought to exclude it (e.g. a freshly 'declined' submission in the 'pending' view) it is omitted from the set of rows displayed. Since `VerdictDialog` is a child of the row itself, the success message is not dropped out of the dom together with it's parent row.

File: client/containers/DashboardSubmissions/VerdictDialog.tsx (hunk 7 above)
Comment (howebenhowe) [reply client/containers/DashboardSubmissions/VerdictDialog.tsx:134]: I am reluctant to break from chained ternary, but I think it's just a matter of taste. Using a `renderInner()` would mean either (a) passing extraneous props into it only to discard them if the necessary component doesn't need them or (b) priming a few components somewhere inside the `VerdictDialog` component before the return value. Those each smell funkier to me.

Here's what converted me a year or so ago:
https://medium.com/javascript-scene/nested-ternaries-are-great-361bddd0f340

Overall, I'm still on the lookout for the best approach to multiple-choice conditional rendering.

File: client/containers/DashboardSubmissions/verdictDialog.scss (hunk 9 above)
Comment (howebenhowe) [reply client/containers/DashboardSubmissions/verdictDialog.scss:10]: `overflow: auto` keeps the formatting bar in view once the email copy runs long

`white-space: break-space` covers for when the placeholder text is longer than the input is wide. This is nearly impossible, and can be removed.

Can't recall why I included `min-height: 100%`, so I'll remove that too.

File: client/containers/DashboardSubmissions/arbitrationMenu.scss (hunk 4 above)
Comment (howebenhowe) [reply client/containers/DashboardSubmissions/arbitrationMenu.scss:2]: I'm happy either way. In the `all` filter view with `grid` the delete / decline / approve buttons stay vertically aligned by effect tolerant of the missing buttons, but at the cost of the goofy state you show above. Comparison, from the figma mock ups:
Screenshot from 2022-01-24 01-55-12

Comment (idreyn) [reply client/containers/DashboardSubmissions/arbitrationMenu.scss:2]: That's a great point! In that case, the "delete" button should definitely go at the end, though, since it's always visible. I think it should also use the `trash` icon, because the `x` is mostly used to mean "Remove from Collection" elsewhere and folks may assume that's the action here, too.

File: client/containers/DashboardSubmissions/VerdictDialog.tsx (hunk 7 above)
Comment (idreyn) [reply client/containers/DashboardSubmissions/VerdictDialog.tsx:134]: Fine by me if you feel strongly about it, though I think the readability argument breaks down when the expressions don't mostly fit onto a single line like this:

```ts
const value =  (!conditionA)
    ? valueC
    : (conditionB)
    ? valueA
    : valueB;
```


 Neither (a) or (b) is necessarily what I'm imagining though. This is a good example of what I mean — the inner React elements are being constructed within the `return` statement, just in an inner function definition.

https://github.com/pubpub/pubpub/blob/a8b1c090a8c523c853d31af520657b630d510b76/client/components/PubShareDialog/PubShareDialog.tsx#L159

Also on (b) — at some point in the last two years @3mcd kind of convinced me that defining React nodes outside of a `return` statement is a totally fine thing to do:

https://github.com/pubpub/pubpub/blob/2965db0ed81ce2f24dd41b59f02cfc0521796edc/client/utils/activity/renderers/itemRenderer.tsx#L28

Just so you know where I am on the carousel of JS code fashions.

File: client/containers/DashboardSubmissions/ArbitrationMenu.tsx [hunk 15]
```diff
@@ -0,0 +1,106 @@
...
+						completedName="deleted"
+						onSubmit={() =>
+							apiFetch('/api/pub', {
+								method: 'DELETE',
+								body: JSON.stringify({
+									id: props.pub.submission.pubId,
+								}),
+							}).then(() => props.pub.submission)
+						}
+						pub={props.pub as DefinitelyHas<Pub, 'submission'>}
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/ArbitrationMenu.tsx:44]: This cast should no longer be necessary, right?

File: client/containers/DashboardSubmissions/ArbitrationMenu.tsx [hunk 16]
```diff
@@ -0,0 +1,106 @@
...
+					<Button
+						minimal
+						small
+						icon={<Icon icon={'cross' as IconName} iconSize={20} />}
+						onClick={openDialog}
+					/>
+				)}
+			>
+				{({ isOpen, onClose }) => (
+					<VerdictDialog
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/ArbitrationMenu.tsx:30]: You might find a `ConfirmDialog` helpful here if you don't want to squeeze this functionality into the `VerdictDialog`. Either way.

File: client/containers/DashboardSubmissions/ArbitrationMenu.tsx [hunk 17]
```diff
@@ -0,0 +1,106 @@
...
+
+const ArbitrationMenu = (props: Props) => (
+	<div className="arbitration-menu-component">
+		<div style={{ gridColumn: 1 }}>
+			<DialogLauncher
+				renderLauncherElement={({ openDialog }) => (
+					<Button
+						minimal
+						small
+						icon={<Icon icon={'cross' as IconName} iconSize={20} />}
```
Comment (idreyn) [inline client/containers/DashboardSubmissions/ArbitrationMenu.tsx:24]: Does this work without the cast?

File: client/containers/DashboardSubmissions/ArbitrationMenu.tsx (hunk 15 above)
Comment (howebenhowe) [reply client/containers/DashboardSubmissions/ArbitrationMenu.tsx:44]: verily

File: client/containers/DashboardSubmissions/ArbitrationMenu.tsx (hunk 17 above)
Comment (howebenhowe) [reply client/containers/DashboardSubmissions/ArbitrationMenu.tsx:24]: Yup.
